votor/src/consensus_pool/vote_pool.rs

```diff
@@ -23,116 +16,100 @@ pub(super) enum VotePool {
     DuplicateBlockVotePool(DuplicateBlockVotePool),
 }
 
+#[derive(Default)]
 pub(super) struct SimpleVotePool {
-    /// Tracks all votes of a specfic vote type made by validators for some slot N.
-    vote_entry: VoteEntry,
-    prev_voted_validators: HashSet<Pubkey>,
+    votes: Vec<VoteMessage>,
+    total_stake: Stake,
+    prev_voted_validators: BTreeSet<Pubkey>,
 }
 
 impl SimpleVotePool {
-    pub(super) fn new() -> Self {
-        Self {
-            vote_entry: VoteEntry::default(),
-            prev_voted_validators: HashSet::new(),
-        }
-    }
-
     pub(super) fn add_vote(
         &mut self,
-        validator_vote_key: &Pubkey,
+        validator_vote_key: Pubkey,
         validator_stake: Stake,
-        vote_message: VoteMessage,
+        vote: VoteMessage,
     ) -> Option<Stake> {
-        if self.prev_voted_validators.contains(validator_vote_key) {
+        if !self.prev_voted_validators.insert(validator_vote_key) {
             return None;
         }
-        self.prev_voted_validators.insert(*validator_vote_key);
-        self.vote_entry.vote_messages.push(vote_message);
-        self.vote_entry.total_stake_by_key = self
-            .vote_entry
-            .total_stake_by_key
-            .saturating_add(validator_stake);
-        Some(self.vote_entry.total_stake_by_key)
+        self.votes.push(vote);
+        self.total_stake = self.total_stake.saturating_add(validator_stake);
+        Some(self.total_stake)
     }
 
-    pub(super) fn add_to_certificate(&self, output: &mut CertificateBuilder) {
-        output
-            .aggregate(&self.vote_entry.vote_messages)
-            .expect("Incoming vote message signatures are assumed to be valid")
+    pub(super) fn votes(&self) -> &[VoteMessage] {
+        &self.votes
     }
 
     pub(super) fn total_stake(&self) -> Stake {
-        self.vote_entry.total_stake_by_key
+        self.total_stake
     }
 
     pub(super) fn has_prev_validator_vote(&self, validator_vote_key: &Pubkey) -> bool {
         self.prev_voted_validators.contains(validator_vote_key)
     }
 }
 
+#[derive(Default)]
+struct VoteEntry {
+    votes: Vec<VoteMessage>,
+    total_stake_by_key: Stake,
+}
+
 pub(super) struct DuplicateBlockVotePool {
     max_entries_per_pubkey: usize,
-    votes: HashMap<Hash, VoteEntry>,
-    total_stake: Stake,
-    prev_voted_block_ids: HashMap<Pubkey, Vec<Hash>>,
+    vote_entries: BTreeMap<Hash, VoteEntry>,
+    prev_voted_block_ids: BTreeMap<Pubkey, BTreeSet<Hash>>,
```

Comment:
I believe we will have up to 3 Hashes here per the protocol, why is BTreeSet better than Vec on 3 elements? (And under most cases this should be 1-element array I think)

Reply:
my thinking is that for a small collection, the performance on a btreeset is comparable to vec and btreeset guarantees at the type level that we do not have duplicates.  It also makes it easier for readers of the code to understand what to expect without inspecting all the call sites to see that we indeed only storing unique elements.